File: c++/031.cpp
class Solution {
/*
1. 找到倒数第1个“顺序对”，记录下“顺序对”的第一个数 nums[k]；
2. [k+1,……,n-1]中找到大于 nums[k]的最小值，并与之交换
3. 对[k+1,n-1]进行插入排序（进阶：此区间定为逆序，即翻转即可）
*/
public:
    void nextPermutation(vector<int>& nums) {
        int n=nums.size(),k=-1,ai;
        for(int i=n-1;i>0;i--){
            if(nums[i]>nums[i-1]){
                k=i-1;
                break;
            }
        }
        if(k>=0){
            ai = k+1;
            for(int i=k+1; i<n; i++){
                if(nums[i]>nums[k]&&nums[i]<nums[ai])
                    ai=i;
            }
            swap(nums[ai],nums[k]);
        }
        for(int i=k+2,j; i<n; i++){
            if(nums[i]<nums[i-1]){
                int t = nums[i];
                for(j=i-1; j>=k+1&&nums[j]>t; j--)
                    nums[j+1] = nums[j];
                nums[j+1] = t; 
            }
        }
    }
};

File: go/031.go
package main

/**
整数数组的下一个排列是指其整数的下一个字典序更大的排列。
给你一个整数数组nums，找出 nums 的下一个排列
必须原地修改，只允许使用额外常数空间
思路：[8,9,6,10,7,2]，从后往前找，找到较小的数6，然后再找到较大的7，交换位置，再将较大的数之后的序列升序。
交换完之后较大数之后的区间为降序区间，只需反转即可。
*/
func nextPermutation(nums []int) {
	k := len(nums) - 1
	for ; k > 0 && nums[k-1] > nums[k]; k-- {
	}
	if k <= 0 {
		reverse(&nums, 0, len(nums)-1)
	} else {
		t := k
		for ; t < len(nums) && nums[k-1] < nums[t]; t++ {
		}
		swap(&nums, k-1, t-1)
		reverse(&nums, k, len(nums)-1)
	}
}
